pkg/controllers/rollout/controller.go

```diff
@@ -0,0 +1,278 @@
+				canBeUnavailableBindings = append(canBeUnavailableBindings, &binding)
+			}
+
+		case fleetv1beta1.BindingStateScheduled:
+			// the scheduler has picked a cluster for this binding
+			schedulerTargetedBinds = append(schedulerTargetedBinds, &binding)
+			// this binding has not been bound yet, so it is an update candidate
+			updateCandidates = append(updateCandidates, &binding)
+
+		case fleetv1beta1.BindingStateBound:
+			schedulerTargetedBinds = append(schedulerTargetedBinds, &binding)
+			canBeReadyBindings = append(canBeReadyBindings, &binding)
+			if isBindingReady(&binding, readyTimeCutOff) {
+				readyBindings = append(readyBindings, &binding)
+			}
+			// The binding needs update if it's not pointing to the latest resource snapshot
+			if binding.Spec.ResourceSnapshotName != latestResourceSnapshotName {
+				removeCandidates = append(removeCandidates, &binding)
+			}
+		}
+	}
+
+	// calculate the target number of bindings
+	targetNumber := 0
+	if crp.Spec.Policy.PlacementType == fleetv1beta1.PickAllPlacementType {
+		// we use the scheduler picked bindings as the target number since there is no target in the CRP
+		targetNumber = len(schedulerTargetedBinds)
+	} else {
+		// the CRP validation webhook should make sure this is not nil
+		targetNumber = int(*crp.Spec.Policy.NumberOfClusters)
+	}
+
+	// calculate the max number of bindings that can be unavailable according to user specified maxUnavailable
+	maxUnavailableNumber, _ := intstr.GetScaledValueFromIntOrPercent(crp.Spec.Strategy.RollingUpdate.MaxUnavailable, targetNumber, true)
+	minAvailableNumber := targetNumber - maxUnavailableNumber
+	// This is the lower bound of the number of bindings that can be available during the rolling update
+	// Since we can't predict the number of bindings that can be unavailable after they are applied, we don't take them into account
+	lowerBoundAvailable := len(readyBindings) - len(canBeUnavailableBindings)
+	for i := 0; i < lowerBoundAvailable-minAvailableNumber; i++ {
+		tobeRemovedBinding = append(tobeRemovedBinding, removeCandidates[i])
+	}
+
+	// calculate the max number of bindings that can be added according to user specified maxUnavailable
+	maxSurgeNumber, _ := intstr.GetScaledValueFromIntOrPercent(crp.Spec.Strategy.RollingUpdate.MaxSurge, targetNumber, true)
+	maxReadyNumber := targetNumber + maxSurgeNumber
+	// This is the upper bound of the number of bindings that can be ready during the rolling update
+	// We count anything that still has work object on the hub cluster as can be ready since the member agent may have connection issue with the hub cluster
+	upperBoundReadyNumber := len(canBeReadyBindings)
+	for i := 0; i < maxReadyNumber-upperBoundReadyNumber; i++ {
+		// we don't differentiate if binding needs to be bound to new cluster or its resource snapshot needs to be updated
+		// TODO:
+		tobeUpgradedBinding = append(tobeUpgradedBinding, updateCandidates[i])
```

Comment:
are we going to differentiate the new bound or snapshot updating in the first phase?

based on the maxSurge definition, we only need to append the new bound ones.